suave/backends/redis_transport.go

```diff
@@ -0,0 +1,183 @@
+package backends
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/flashbots/go-utils/cli"
+	"github.com/go-redis/redis/v8"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	suave "github.com/ethereum/go-ethereum/suave/core"
+)
+
+var (
+	redisUpsertTopic = "store:upsert"
+
+	redisConnectionPoolSize = cli.GetEnvInt("REDIS_CONNECTION_POOL_SIZE", 0) // 0 means use default (10 per CPU)
+	redisMinIdleConnections = cli.GetEnvInt("REDIS_MIN_IDLE_CONNECTIONS", 0) // 0 means use default
+	redisReadTimeoutSec     = cli.GetEnvInt("REDIS_READ_TIMEOUT_SEC", 0)     // 0 means use default (3 sec)
+	redisPoolTimeoutSec     = cli.GetEnvInt("REDIS_POOL_TIMEOUT_SEC", 0)     // 0 means use default (ReadTimeout + 1 sec)
+	redisWriteTimeoutSec    = cli.GetEnvInt("REDIS_WRITE_TIMEOUT_SEC", 0)    // 0 means use default (3 seconds)
+
+	formatRedisBidKey = func(bidId suave.BidId) string {
+		return fmt.Sprintf("bid-%x", bidId)
+	}
+
+	formatRedisBidValueKey = func(bidId suave.BidId, key string) string {
+		return fmt.Sprintf("bid-data-%x-%s", bidId, key) // TODO: should also include the hash of the bid at least
+	}
+)
+
+type RedisPubSubTransport struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	redisUri string
+	client   *redis.Client
+}
+
+func NewRedisPubSubTransport(redisUri string) *RedisPubSubTransport {
+	return &RedisPubSubTransport{
+		redisUri: redisUri,
+	}
+}
+
+func (r *RedisPubSubTransport) Start() error {
+	if r.cancel != nil {
+		r.cancel()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.ctx = ctx
+	r.cancel = cancel
+
+	client, err := connectRedis(r.redisUri)
+	if err != nil {
+		return err
+	}
+	r.client = client
+
+	return nil
+}
+
+func (r *RedisPubSubTransport) Stop() error {
+	if r.cancel == nil || r.client == nil {
+		panic("Stop() called before Start()")
```

Comment:
no panic please

Reply:
For this one I'm kind of on the fence. This should never be the case, and would only appear if someone introduced a bug - in which case a panic is the quickest way to spot the error.
On the other hand it's not critical so it'd be better to not terminate.
I'll return errors